File: tests/typesafety/python/invoice_typestate_misuse.py
# expect: fails
# title: Paying a draft that was never sent, rejected by the checker
# misuse-of: invoice_typestate
from dataclasses import dataclass


@dataclass(frozen=True)
class DraftInvoice:
    number: str

    def send(self) -> "SentInvoice":
        return SentInvoice(self.number)


@dataclass(frozen=True)
class SentInvoice:
    number: str

    def record_payment(self) -> "PaidInvoice":
        return PaidInvoice(self.number)


@dataclass(frozen=True)
class PaidInvoice:
    number: str


paid = DraftInvoice("INV-7").record_payment()

File: tests/typesafety/typescript/invoice_typestate_misuse.ts
// expect: fails

type DraftInvoice = { readonly stage: "draft"; readonly number: string };
type SentInvoice = { readonly stage: "sent"; readonly number: string };
type PaidInvoice = { readonly stage: "paid"; readonly number: string };

function send(invoice: DraftInvoice): SentInvoice {
  return { stage: "sent", number: invoice.number };
}

function recordPayment(invoice: SentInvoice): PaidInvoice {
  return { stage: "paid", number: invoice.number };
}

export const paid = recordPayment({ stage: "draft", number: "INV-7" });
